### ISSUE TITLE
HTTP/2 to upstream

### ISSUE THREAD

**Issue by PrabhatDixit** [nginx/nginx#395]

**Comment by DemiMarie** [comment]
This is already implemented in Angie.

**Comment by HanadaLee** [comment]
> This is already implemented in Angie.

If I remember correctly, Angie only implemented HTTP/3 to Upstream, and there is no implementation of H2 to Upstream.

There was a patch to implement this feature on the nginx development mailing list. It was provided by Google. However, it was shelved because of concerns that implementing two different protocols in the same module (proxy) violated the nginx design principles.

**Comment by DemiMarie** [comment]
One way to implement this would be:

1. Add HTTP/2 client support to the existing HTTP/2 code.
2. Have the new module (`proxy_http2`?) depend on this library code.
3. Have support in the core `proxy` module for multiple HTTP modules, chosen based on ALPN negotiation.

That said, does following the nginx design principles really provide any benefit here?

**Comment by HanadaLee** [comment]
> One way to implement this would be:
> 
> 1. Add HTTP/2 client support to the existing HTTP/2 code.
> 2. Have the new module (`proxy_http2`?) depend on this library code.
> 3. Have support in the core `proxy` module for multiple HTTP modules, chosen based on ALPN negotiation.
> 
> That said, does following the nginx design principles really provide any benefit here?

Yes, it causes a lot of code duplication, and support for alpn configuration becomes impossible. The grpc module is an example. The grpc module is actually a proxy_http2 module (except for some customized processing to cater to the grpc protocol, and the cancellation of caching and other functions). I am now using the proxy and grpc modules at the same time, which requires a lot of repeated configuration.



https://github.com/arut/nginx/tree/http3-upstream
Judging from the code of arut's http 3-upstream branch, they now seem to be inclined to complete the three http protocols in the proxy module.

**Comment by DemiMarie** [comment]
It definitely seems to me that having HTTP/1.x, HTTP/2, and HTTP/3 be unified is the better approach to me.  The only exception would be if one wanted to allow third party modules to provide their own HTTP implementations, and I am not sure if that is a use-case NGINX should support.

**Comment by aronmolnar** [comment]
Pushing this due to security reasons: https://http1mustdie.com/

**Comment by dmitry-j-mikhin** [comment]
https://github.com/nginx/nginx/pull/771 - merged
Announced in 1.29.4 - https://github.com/nginx/nginx/pull/1023
